src/NonLinearProgram/nlp_utilities.jl

```diff
@@ -0,0 +1,556 @@
+
+Build the M (KKT Jacobian w.r.t. solution) and N (KKT Jacobian w.r.t. parameters) matrices for the sensitivity analysis.
+"""
+function build_M_N(
+    model::Model,
+    cons::Vector{MOI.Nonlinear.ConstraintIndex},
+    _X::AbstractVector,
+    _V_L::AbstractVector,
+    _X_L::AbstractVector,
+    _V_U::AbstractVector,
+    _X_U::AbstractVector,
+    leq_locations::Vector{Z},
+    geq_locations::Vector{Z},
+    ineq_locations::Vector{Z},
+    has_up::Vector{Z},
+    has_low::Vector{Z},
+) where {Z<:Integer}
+    # Setting
+    num_vars = get_num_primal_vars(model)
+    num_parms = get_num_params(model)
+    num_cons = get_num_constraints(model)
+    num_ineq = length(ineq_locations)
+    num_low = length(has_low)
+    num_up = length(has_up)
+
+    # Primal solution
+    X_lb = spzeros(num_low, num_low)
+    X_ub = spzeros(num_up, num_up)
+    V_L = spzeros(num_low, num_vars + num_ineq)
+    V_U = spzeros(num_up, num_vars + num_ineq)
+    I_L = spzeros(num_vars + num_ineq, num_low)
+    I_U = spzeros(num_vars + num_ineq, num_up)
+
+    # value and dual of the lower bounds
+    for (i, j) in enumerate(has_low)
+        V_L[i, j] = _V_L[j]
+        X_lb[i, i] = _X[j] - _X_L[j]
+        I_L[j, i] = -1
+    end
+    # value and dual of the upper bounds
+    for (i, j) in enumerate(has_up)
+        V_U[i, j] = _V_U[j]
+        X_ub[i, i] = _X_U[j] - _X[j]
+        I_U[j, i] = 1
+    end
+
+    # Function Derivatives
+    hessian, jacobian = compute_optimal_hess_jac(model, cons)
+    primal_idx = [i.value for i in model.cache.primal_vars]
+    params_idx = [i.value for i in model.cache.params]
+    # Hessian of the lagrangian wrt the primal variables
+    W = spzeros(num_vars + num_ineq, num_vars + num_ineq)
+    W[1:num_vars, 1:num_vars] = hessian[primal_idx, primal_idx]
+    # Jacobian of the constraints
+    A = spzeros(num_cons, num_vars + num_ineq)
+    # A is the Jacobian of: c(x) = b and c(x) <= b and c(x) >= b, possibly all mixed up.
+    # Each of the will be re-written as:
+    # c(x) - b = 0
+    # c(x) - b - su = 0, su <= 0
+    # c(x) - b - sl = 0, sl >= 0
+    # Jacobian of the constraints wrt the primal variables
+    A[:, 1:num_vars] = jacobian[:, primal_idx]
+    # Jacobian of the constraints wrt the slack variables
+    for (i, j) in enumerate(geq_locations)
+        A[j, num_vars+i] = -1
+    end
+    for (i, j) in enumerate(leq_locations)
+        A[j, num_vars+i] = -1
+    end
+    # Partial second derivative of the lagrangian wrt primal solution and parameters
+    ∇ₓₚL = spzeros(num_vars + num_ineq, num_parms)
+    ∇ₓₚL[1:num_vars, :] = hessian[primal_idx, params_idx]
+    # Partial derivative of the equality constraintswith wrt parameters
+    ∇ₚC = jacobian[:, params_idx]
+
+    # M matrix
+    # M = [
+    #     [W A' -I I];
+    #     [A 0 0 0];
+    #     [V_L 0 (X - X_L) 0]
+    #     [V_U 0 0 0 (X_U - X)]
+    # ]
+    len_w = num_vars + num_ineq
+    M = spzeros(
```

Comment:
I think there should be a cleaner way to build `M` and `N`, with less allocations. Let me try to build a MWE, we can assess later if we should address this in this PR or later.